### DISCUSSION TITLE
`FormatException` when converting pandas.DataFrame to iterable quotes

### DISCUSSION THREAD

**Discussion by masoudaz89** [DaveSkender/Stock.Indicators#1164] [Help and support]
FormatException: The input string '42.84749984741211' was not in a correct format.
   at System.Number.ThrowFormatException[TChar](ReadOnlySpan`1 value)
   at System.Number.ParseDecimal[TChar](ReadOnlySpan`1 value, NumberStyles styles, NumberFormatInfo info)
   at System.Decimal.Parse(String s)
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeDirectByRefWithFewArgs(Object obj, Span`1 copyOfArgs, BindingFlags invokeAttr)

**Comment by DaveSkender** [comment]
Can you share a bit more about your setups that led to this error? Ideally, a code snippet of how you're converting to iterable Quotes?  Here's an example, if helpful: [Williams Fractal with DataFrames conversion](https://replit.com/@daveskender/Stock-Indicators-for-Python-Williams-Fractal)

More info to consider:

- #1165

**Reply by masoudaz89** [reply]
from stock_indicators import Quote



quotes_list = [
    Quote(d,o,h,l,c,v) 
    for d,o,h,l,c,v 
    in zip(df.index, df['Open'], df['High'], df['Low'], df['Close'], df['Volume'])
]
this is my code for converting to iterable Quotes 
I download data from yahoo finance in data frame mode

**Reply by DaveSkender** [reply]
is `df.index` something that can be converted to `datetime` format?  I’d expect that to be something like `df['Date']`

**Reply by masoudaz89** [reply]
yes thats ok its not the problem

**Reply by DaveSkender** [reply]
Also, check to ensure you’re not adding extra `‘` in the data frame values, `'42.847…`

**Reply by masoudaz89** [reply]
ok

**Reply by masoudaz89** [reply]
I checked the data and it was okay. All data types were float, except for the datetime.

**Reply by DaveSkender** [reply]
If you upload your raw (pre-converted) quotes data, I'll try to reproduce over the next few days.
From the information you've provided, I can't see why or when this error would happen with the value you've shown.

There's really only two reasons it'd throw a [FormatException](https://learn.microsoft.com/en-us/dotnet/api/system.convert.todecimal?view=net-8.0#system-convert-todecimal(system-string)):

1. The string is not in a decimalized format that is recognizable (e.g. bad characters)
2. The local system current culture info is mismatched with the data type (e.g. commas instead of decimal points, possibly)

The error you've shown above implies that it's presenting as a `string` value, but that could possibly be how Python handles certain conversions under the hood, despite what you've provided.

**Reply by masoudaz89** [reply]
https://github.com/masoudaz89/indicator.git
i uploaded data and code
you can check them both

**Reply by DaveSkender** [reply]
what version do you see when you run `dotnet --version` in your terminal?

**Reply by DaveSkender** [reply]
I'm not able to reproduce directly with the provided CSV quotes, but I did accidentally discover something that might point to the problem here.  But first, this works fine ([fork and run on Replit](https://replit.com/@daveskender/Stock-Indicators-for-Python-Williams-Fractal-Issue-1164))

```python
import pandas as pd
from stock_indicators import indicators, Quote

# to demonstrate conversion from DataFrames to Quotes
df = pd.read_csv('https://raw.githubusercontent.com/masoudaz89/indicator/main/dataset.csv', parse_dates=['Date'])

# convert to iterable Quotes
quotes = [ 
  Quote(date, open, high, low, close, volume) 
    for date, open, high, low, close, volume 
    in zip(df['Date'],
           df['Open'], 
           df['High'], 
           df['Low'],
           df['Close'],
           # df['Adj Close'],
           df['Volume'], strict=True)]

# show the iterable Quotes
print("Date | Open | High | Low | Close | Volume")
for q in quotes: 
  print(f"{q.date:%Y-%m-%d} | {q.open} | {q.high} | {q.low} | {q.close} | {q.volume}")

# calculate indicators
results = indicators.get_fractal(quotes)

# show the indicator results
print("Date       | Bear   | Bull")
for r in results: 
  print(f"{r.date:%Y-%m-%d} | {r.fractal_bear or '':6} | {r.fractal_bull or ''}")
```

Side note: you'll probably want to use `Adj Close` from the Yahoo! feed.  The `Close` value is not adjusted for splits or dividends.

**Reply by DaveSkender** [reply]
What I accidentally found was about _sourcing_ your CSV data.
When putting the URL in for the CSV (above), I first accidentally used `https://github.com/masoudaz89/indicator/blob/main/dataset.csv` and got the `FormatException` error ... because it was trying to load the actual HTML page. 

Then used the `raw` file URL `https://raw.githubusercontent.com/masoudaz89/indicator/main/dataset.csv` and it worked fine.

So, check the method you're using to fetch this data, maybe there's a bug in that tool.  While the CSV you provided is clean, I suspect that's just a raw download from the Yahoo website and not actually what your code is getting.

Either that or there's some environmental factors in your dev environment that I can't really test without more info.
Is your dev operating environment setup to use commas instead of decimals in numeric formats?  (shouldn't matter since we use Invariant culture settings, but I'd not be surprised if something somewhere doesn't like it).

**Reply by LeeDongGeon1996** [reply]
@DaveSkender I suspect that we don't set `CultureInfo` when parse decimal string to C# Decimal object. Because, in the original post, `FormatException: The input string '42.84749984741211' was not in a correct format.` doesn't make sense. '42.84749984741211' is just super normal decimal string.

` ...I'm looking further into it now. ...`

**Reply by DaveSkender** [reply]
I agree, I think this is caused by the _string_ format possible cause I'd [enumerated above](https://github.com/DaveSkender/Stock.Indicators/discussions/1164#discussioncomment-8459302).  The underlying .NET code doesn't set or expect a specific culture variant, nor does it set an _invariant_ culture that also defaults to use of decimals -- the library _should_ simply operate entirely based on its environment settings without having to do anything special.

**Reply by LeeDongGeon1996** [reply]
@DaveSkender Right. But in Python implementation, we have a conversion from decimal string to C# Decimal object. It might need a CultureInfo. 
Is `CultureInfo.InvariantCulture` for the general usage, ignoring user's local `CultureInfo`?

**Reply by DaveSkender** [reply]
My theory is that as long as users provide dates and numeric values that match their environment, there shouldn't be a problem.  Forcing the use of `InvariantCulture` will _force a requirement to use decimal points_ instead of the locally set culture, so that's not likely a solution.  Let me do some .NET Fiddle to try out some scenarios.  We may also need to add some unit tests in both packages to isolate if there's an actual code problem on our end, or to confirm my theory.  It could just be a mismatched user environment setup, which we have no control over.

**Reply by LeeDongGeon1996** [reply]
@DaveSkender I've tested on all `CultureInfo`s with the codes below
```
import stock_indicators
from System.Globalization import CultureInfo, CultureTypes
from stock_indicators._cslib import CsDecimal

test_num_string= '0.50000000'

incompatible_culture_info = []
all_cultures = CultureInfo.GetCultures(CultureTypes.AllCultures)
for c in all_cultures:
    try:
        print(f'{CsDecimal.Parse(test_num_string, c)} {c.Name} {c.EnglishName}')
    except:
        incompatible_culture_info.append(c.Name)

print(incompatible_culture_info)
```

And found all incompatible culture with normal (at least in my culture) decimal string, '0.50000000'. You can see many cultures are incompatible with that number. Furthermore, there are even some cultures, for example `tr (Turkish)`, which converts '0.50000000' into '50000000'.

**Reply by LeeDongGeon1996** [reply]
can be quite tricky. Because we don't know which locale of data user has.
 - https://herrmann.tech/en/blog/2021/02/05/how-to-deal-with-international-data-formats-in-python.html
 - https://stackoverflow.com/questions/3425294/how-to-detect-the-os-default-language-in-python

**Reply by DaveSkender** [reply]
so that’s a test of how each culture handles decimals, which in itself isn’t necessary something to fix in our library.

If a user is working on a machine with a different environment culture from that of the quotes, and trying to import data that conflict with it, it’s up to them to notate the culture _of the quotes_ OR override locale settings in their code to match the quotes.  

What’s probably confusing people is that [DataFrame doesn’t use locale culture](https://pandas.pydata.org/docs/reference/api/pandas.io.formats.style.Styler.format.html#pandas.io.formats.style.Styler.format) info unless specified, but .NET does, which is why it’s breaking on conversion.

They may simply need to set [locale](https://docs.python.org/3.8/library/locale.html) before converting  `locale.setlocale(locale.LC_ALL, 'C')` (not sure I picked the right syntax) or as a parameter in the converter they’re using noting quotes culture, but only when they’re incompatible.  So, this is correct (below), but applied by the user, not in our library code … or the Python equivalent `setlocale`

> `CultureInfo.InvariantCulture` for the general usage, ignoring user's local

**Reply by DaveSkender** [reply]
On a meta level, this is a conflict over whether you _honor or allow_ use of locale culture standards or force them to use international or _western_ standards.   I prefer to honor them, so am not a fan forcing one model, but may be open to an international standard default — since most quotes sources are in that format — as long as it can be overridden and set to local, like they do with DataFrame.  For now, I think users will need to make an explicit choice.

**Comment by DaveSkender** [comment]
Right now we’re using this string converter:
https://github.com/DaveSkender/Stock.Indicators.Python/blob/88c887b09782cd7614286944add6ac6bc2b104af/stock_indicators/_cstypes/decimal.py#L21

Maybe what we can do here is have this failover to try [a second method](https://learn.microsoft.com/en-us/dotnet/api/system.decimal.parse?view=net-8.0#system-decimal-parse(system-string-system-globalization-numberstyles)) to make it more resilient when user environments don’t match source quote formats.  It’d still take the local culture first, but then try a more standard one

```Python
return CsDecimal.Parse(str(decimal), AllowDecimalPoint)
```

**Reply by LeeDongGeon1996** [reply]
@DaveSkender I agree with adding failover step, but am worried about `AllowDecimalPoint` gets decimal separator from `CultureInfo`. So.. why don't we set western standard as failover method. 
>Otherwise, the decimal separator character is determined by the [NumberDecimalSeparator](https://learn.microsoft.com/en-us/dotnet/api/system.globalization.numberformatinfo.numberdecimalseparator?view=net-8.0#system-globalization-numberformatinfo-numberdecimalseparator) property.
https://learn.microsoft.com/en-us/dotnet/api/system.globalization.numberstyles?view=net-8.0

----
Also, even some cultures convert completely wrong(`0.5` -> `5.0`) but don't throw errors as mentioned above. <- ~right,, this can not be controlled on our side. user should be careful about this.~
> Furthermore, there are even some cultures, for example tr (Turkish), which converts '0.50000000' into '50000000'.

**Reply by DaveSkender** [reply]
I may not have the exactly correct syntax.  General _invariant_ culture might work for the failover too.  I don't think we should try to address every possible cultural variation, like the Turkish one -- we'd never really know if their source quotes were in an aligned culture and overriding it would _be bad and irreversible_ if we did.  At some point, the user will need to specify culture info appropriately themselves.

**Reply by DaveSkender** [reply]
We should probably just start with some guidance on _how to convert quotes when your local culture settings are incompatible_ and see if that's sufficient.  While the failover could help, I do see that it isn't a perfect solution either.  Makes you wonder how DataFrame handles import of non-standard culture data; does it fail or does it convert them all on import?

**Reply by LeeDongGeon1996** [reply]
@DaveSkender 
> does it fail or does it convert them all on import?

After a quick search, It looks like it'd be failed, if users input wrong(`"wrong" means locale is not matched with their LC_NUMERIC`) format of data. The belows are just stackoverflow though, but anyway can be reffered to as some examples:
 - https://stackoverflow.com/questions/22137723/convert-number-strings-with-commas-in-pandas-dataframe-to-float
 - https://stackoverflow.com/questions/57575535/trying-to-use-locale-atof-for-changing-the-period-in-decimals-within-a-datafra
 - https://stackoverflow.com/questions/53889860/display-pandas-dataframe-with-european-formatting 

And I could have found some usecases on above. Usually, they convert their numeric data into their desired format, before they input to DataFrame. The most of users who have submitted ISSUEs to us have data in International format, which uses `.(comma)` as a separator. It means the problem is `locale` setting on their machine. And I don't want to override their locale setting with `setlocale()`, because it can also affect to unexpected somthing.

> We should probably just start with some guidance on how to convert quotes when your local culture settings are incompatible and see if that's sufficient.

As a result, YES. totally agree with it. We can also add failover step on our [decimal.py](https://github.com/DaveSkender/Stock.Indicators.Python/blob/main/stock_indicators/_cstypes/decimal.py), I'm worried about that could have a negative impact on performance. so let's just start with guiding how to convert their data.

**Reply by LeeDongGeon1996** [reply]
Also, we can add some explanatory error message when `FormatException` is thrown.

**Reply by LeeDongGeon1996** [reply]
The patched version has been released in [v1.3.1](https://github.com/facioquo/stock-indicators-python/releases/tag/1.3.1).

**Comment by LeeDongGeon1996** [comment]
@DaveSkender
I think there are two ways after adding error message:
1. Add guide(FAQ) page in docs. And add that link in error message (https://github.com/DaveSkender/Stock.Indicators.Python/pull/351)
2. Just tell users how to resolve this problem every time they create issue.

How do you think?

**Reply by DaveSkender** [reply]
Yes, having something in the guide is a good idea, and a more helpful error message is always a good to do.

> Just tell users how to resolve this problem every time they create issue.

Well, if we have to, but let's have some good docs in the guide to make it easier to direct them to (and possibly a cleaned up #1165)

**Reply by LeeDongGeon1996** [reply]
Yeah, agree with you.
Now I'm thinking about how/where we can compose this guide well, in terms of docs site UX

**Reply by DaveSkender** [reply]
Since `quotes` are really the only place we see this, I'd suggest somewhere in the https://python.stockindicators.dev/guide/#historical-quote section.  It'd only need to be a short blurb, possibly with a link to #1165 for more _troubleshooting_ information.

**Reply by LeeDongGeon1996** [reply]
I'll try to make a draft.
